### PR TITLE
[S3 API] Remove Ambry header from response

### PR DESCRIPTION
**Pull request by weisong44**
Changes introduced in this PR are
* Introduced base class S3BaseHandler, which allows pre- and post-processing
* Added post-processing to remove Ambry header from response right before invoking the final callback
* Some basic erroring checking and logging

### CODE REVIEW

**Review by JingQianCloud — APPROVED** [review]
I like the idea to have the base S3 class.
Code looks good to me.
Let's see if Arun has any opinion if the S3 base class fits all his S3 handlers.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BaseHandler.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+   * @param restRequest the {@link RestRequest} that contains the request headers and body.
+   * @param restResponseChannel the {@link RestResponseChannel} that contains the response headers and body.
+   * @param callback the {@link Callback} to invoke when the response is ready (or if there is an exception).
+   * @throws RestServiceException exception when the processing fails
+   */
+  public void handle(RestRequest restRequest, RestResponseChannel restResponseChannel,
+      Callback<R> callback) throws RestServiceException {
+    String path = ((RequestPath) restRequest.getArgs().get(REQUEST_PATH)).getOperationOrBlobId(true);
+    if (!path.startsWith(Operations.NAMED_BLOB)) {
+      throw new RuntimeException("S3 request handler can only handle named blob requests");
```
Comment (Arun-LinkedIn) [inline ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BaseHandler.java:65]: Just a question. Would it be better to send `400 bad request` to user by completing the callback with `new RestServiceException("S3 request handler can only handle named blob requests", RestServiceErrorCode.BadRequest` ? I think `RuntimeException` translates to `500 internal server error` to user.

**Review by Arun-LinkedIn — APPROVED** [review]
Looks good to me. Thank you for introducing the base class :) Just had one minor question.

File: ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BaseHandler.java (hunk 1 above)
Comment (weisong44) [reply ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BaseHandler.java:65]: My understanding is that this should never happen, i.e. we internally translate /s3 to /named and do other checks before we let the request proceed, so I chose 500. However, if there is a chance for client to construct to land here, then I agree that we should return 400.

**Pull request merged**